File: backend/src/main/java/cn/etarch/mao/session/util/TitleGenerator.java
package cn.etarch.mao.session.util;

import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class TitleGenerator {

    private static final int MAX_TITLE_LENGTH = 50;
    private static final Pattern SKILL_PATTERN = Pattern.compile("\\$\\{([^}]+)\\}\\$");
    private static final Pattern COMMAND_PATTERN = Pattern.compile("#\\{([^}]+)\\}#");

    /**
     * Preprocess user message for title generation:
     * - Skill markers (${name}$): stripped from mixed text; converted to "/name" if sole content
     * - Command markers (#{name}#): expanded to their content
     */
    public static String preprocessForTitle(String text, Map<String, String> commandContentMap) {
        if (text == null || text.isBlank()) {
            return text;
        }

        String result = text;

        // Check if text is ONLY a single skill marker with no other content
        Matcher soleSkillMatcher = SKILL_PATTERN.matcher(result.trim());
        if (soleSkillMatcher.matches()) {
            return "/" + soleSkillMatcher.group(1);
        }

        // Strip all skill markers from mixed content
        result = SKILL_PATTERN.matcher(result).replaceAll("");

        // Expand command markers: #{name}# → content
        if (commandContentMap != null && !commandContentMap.isEmpty()) {
            Matcher cmdMatcher = COMMAND_PATTERN.matcher(result);
            StringBuffer sb = new StringBuffer();
            while (cmdMatcher.find()) {
                String cmdName = cmdMatcher.group(1);
                String replacement = commandContentMap.getOrDefault(cmdName, cmdMatcher.group(0));
                cmdMatcher.appendReplacement(sb, Matcher.quoteReplacement(replacement));
            }
            cmdMatcher.appendTail(sb);
            result = sb.toString();
        }

        return result.trim();
    }

    public static String generate(String userMessage) {
        if (userMessage == null || userMessage.isBlank()) {
            return null;
        }

        String trimmed = userMessage.trim();

        // Use user message as-is, truncated
        return truncate(trimmed, MAX_TITLE_LENGTH);
    }

    private static String truncate(String text, int maxLength) {
        if (text == null) return "";
        if (text.length() <= maxLength) return text;
        return text.substring(0, maxLength) + "...";
    }
}

File: backend-ts/src/session/util/title-generator.ts
const MAX_TITLE_LENGTH = 50;
const SKILL_PATTERN = /\$\{([^}]+)\}\$/g;
const COMMAND_PATTERN = /#\{([^}]+)\}#/g;

export function preprocessForTitle(text: string | null | undefined, commandContentMap: Map<string, string> | Record<string, string> | null): string | null | undefined {
  if (text == null || text.trim().length === 0) {
    return text;
  }

  let result = text;
  const trimmed = result.trim();
  const sole = /^\$\{([^}]+)\}\$/.exec(trimmed);
  if (sole && sole[0] === trimmed) {
    return `/${sole[1]}`;
  }

  result = result.replace(SKILL_PATTERN, '');

  const map = toMap(commandContentMap);
  if (map.size > 0) {
    result = result.replace(COMMAND_PATTERN, (full, name: string) => map.get(name) ?? full);
  }

  return result.trim();
}

export function generate(userMessage: string | null | undefined): string | null {
  if (userMessage == null || userMessage.trim().length === 0) {
    return null;
  }
  return truncate(userMessage.trim(), MAX_TITLE_LENGTH);
}

function truncate(text: string, maxLength: number): string {
  if (text.length <= maxLength) {
    return text;
  }
  return `${text.slice(0, maxLength)}...`;
}

function toMap(src: Map<string, string> | Record<string, string> | null | undefined): Map<string, string> {
  if (!src) {
    return new Map();
  }
  if (src instanceof Map) {
    return src;
  }
  return new Map(Object.entries(src));
}

export const TitleGenerator = { preprocessForTitle, generate };
